### PR TITLE
feat: add FlashInfer Trace Apply

### PR DESCRIPTION
**Pull request by yongwww**
## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Reviewer Notes

Co-authored-by: Yixin Dong <yixind@andrew.cmu.edu>
Co-authored-by: Zihao Ye <zihaoy@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Opt-in "Trace Apply" runtime to dispatch select API calls to curated solutions; enable via env var or API.
  * Solution loaders for Python and compiled backends; per-solution caching and safe materialization.
  * Stateful plan/run support: capture/reuse plan-time inputs for run-time dispatch.

* **Behavior Changes**
  * Strict dispatch semantics: matched solutions execute (errors propagate); unmatched calls fall back.
  * Templates: outputs can bind to API params; several attention templates now optionally return LSE.

* **Documentation**
  * Added comprehensive Trace Apply guide.

* **Tests**
  * Extensive unit and integration tests for Trace Apply and loaders.

### CODE REVIEW

**Review by nv-yunzheq — APPROVED** [review]

File: flashinfer/trace/solution.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    """Languages a solution implementation may be written in."""
+
+    PYTHON = "python"
+    TRITON = "triton"
+    CUTEDSL = "cutedsl"
+    TILELANG = "tilelang"
+    CPP = "cpp"
+    CUDA = "cuda"
+    CUTLASS = "cutlass"
+    CUTILE = "cutile"
```
Comment (nv-yunzheq) [inline flashinfer/trace/solution.py:43]: cutile should be most likely python, right?

File: docs/trace_apply.rst [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+The folder layout
+-----------------
+
+``FLASHINFER_TRACE_PATH`` points at a directory in the flashinfer-bench layout::
+
+    <root>/
+      definitions/**/*.json    # problem specs: axes, inputs (incl. dtype), fi_api tag
+      solutions/**/*.json       # implementations — one solution per file
+      workloads/ traces/ ...    # ignored by apply
```
Comment (nv-yunzheq) [inline docs/trace_apply.rst]: What is the kernel only support selected problem sizes of certain definition? Would it mean apply will fail?
For example, if I have gemm only works on M=1,2,4,8. N=128, K=128. While the definition only constrain N,K to be 128, does not set constrain on M. Would it fail?

File: docs/trace_apply.rst [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+``solutions/`` provides the kernels to substitute; ``definitions/`` provides the
+routing identity. Other subdirectories are ignored. A folder may hold many
+solutions, and **all of them are applied at once** — one kernel substituted per
```
Comment (nv-yunzheq) [inline docs/trace_apply.rst]: Could you elaborate a bit? What do you mean by apply all the solution, but only one kernel per definition?

File: flashinfer/trace_apply/adapt.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+from typing import Any, Callable
+
+import torch
+
+# OUTPUT_DESTS maps, per fi_api, each Definition output -> the API parameter
+# holding its destination buffer (an input tensor for in-place APIs; ``out=`` /
+# ``lse`` for value-returning ones). OUTPUT_ACTIVATION gates optional outputs by
+# a runtime flag (``lse`` is active only when ``return_lse=True``) so the
+# returned arity matches the API.
+OUTPUT_DESTS: dict[str, dict[str, str]] = {
```
Comment (nv-yunzheq) [inline flashinfer/trace_apply/adapt.py]: Do we plan to maintain this, or is there a way to better align with flashinfer trace and function development

File: tests/trace_apply/test_trace_apply_install.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+SENTINEL = 4242.0
+# rmsnorm inputs are 'hidden_states' (param 'input') and 'weight', both float32.
+DT_F32 = frozenset({("hidden_states", "float32"), ("weight", "float32")})
+
+
+def _write_trace(root: Path) -> None:
+    (root / "definitions" / "rmsnorm").mkdir(parents=True, exist_ok=True)
+    (root / "solutions" / "rmsnorm").mkdir(parents=True, exist_ok=True)
+
+    (root / "definitions" / "rmsnorm" / "rmsnorm_h16.json").write_text(
```
Comment (nv-yunzheq) [inline tests/trace_apply/test_trace_apply_install.py]: Does it mean if we want to test it out, it would write things on the fly? And it seems to be a fake function. Could you through something runnable kernel into solution folder and test it out?

File: flashinfer/trace/solution.py (hunk 1 above)
Comment (yongwww) [reply flashinfer/trace/solution.py:43]: yes, have added CUTILE to _PYTHON_FAMILY

File: docs/trace_apply.rst (hunk 2 above)
Comment (yongwww) [reply docs/trace_apply.rst]: M is a variable axis, so it's not part of the definition identity — one solution serves all M. Trace Apply will raise error if the kernel can't handle a given M

File: docs/trace_apply.rst (hunk 3 above)
Comment (yongwww) [reply docs/trace_apply.rst]: Old wording. It's now enable_apply({definition_name: solution}) ; each definition maps to a single kernel.

File: flashinfer/trace_apply/adapt.py (hunk 4 above)
Comment (yongwww) [reply flashinfer/trace_apply/adapt.py]: Good point — removed the hardcoded table. Output bindings now come from the trace template's param field.

File: tests/trace_apply/test_trace_apply_install.py (hunk 5 above)
Comment (yongwww) [reply tests/trace_apply/test_trace_apply_install.py]: Done — added test_enable_apply_real_kernel_numeric (real RMSNorm checked against the reference) plus a real-callable dispatch test
